src/theory/uf/eq_proof.cpp

```diff
@@ -1454,12 +1455,50 @@ Node EqProof::addToProof(CDProof* p,
   // rewriting
   if (!CDProof::isSame(conclusion, d_node))
   {
-    Trace("eqproof-conv") << "EqProof::addToProof: add "
+    Trace("eqproof-conv") << "EqProof::addToProof: try to flatten via a"
                           << PfRule::MACRO_SR_PRED_TRANSFORM
-                          << " step to flatten rebuilt conclusion "
-                          << conclusion << "into " << d_node << "\n";
-    p->addStep(
-        d_node, PfRule::MACRO_SR_PRED_TRANSFORM, {conclusion}, {d_node}, true);
+                          << " step the rebuilt conclusion "
+                          << conclusion << " into " << d_node << "\n";
+    Node res = p->getManager()->getChecker()->checkDebug(
+        PfRule::MACRO_SR_PRED_TRANSFORM,
+        {conclusion},
+        {d_node},
+        Node::null(),
+        "eqproof-conv");
+    // If rewriting was not able to flatten the rebuilt conclusion into the
+    // original one, we give up and use a TRUST_REWRITE step, generating a proof
+    // for the original conclusion d_node such as
+    //
+    //     Converted EqProof
+    //  ----------------------         ----------------------- TRUST_REWRITE
+    //     conclusion                    conclusion = d_node
+    // ------------------------------------------------------- EQ_RESOLVE
+    //                       d_node
+    //
+    //
+    //  If rewriting was able to do it, however, we just add the macro step.
+    if (res.isNull())
+    {
+      Trace("eqproof-conv")
+          << "EqProof::addToProof: adding a trust rewrite step\n";
+      Node bridgeEq = conclusion.eqNode(d_node);
+      p->addStep(bridgeEq,
+                 PfRule::TRUST_REWRITE,
```

Comment:
Can you use a new trusted PfRule for this?  Trusted proof rules should ideally be 1-to-1 from call sites in the code so that the source of these kinds of failures are easy to track.

Reply:
Sounds good. I've added one now.